File: python/server/routes/friends.py
from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from ..database import get_db
from ..models.user import User
from ..models.friend import Friend
from ..schemas.friend import FriendRequest
from ..services.friend_service import (
    send_friend_request,
    respond_friend_request,
    remove_friend,
)
from ..utils import decode_token

router = APIRouter(prefix="/api/friends")


async def get_user_id(request: Request):
    auth = request.headers.get("authorization", "")
    payload = decode_token(auth)
    if not payload:
        raise HTTPException(status_code=401, detail="Unauthorized")
    return int(payload["sub"])


@router.post("/friend_request")
async def friend_request(body: FriendRequest, request: Request, session: AsyncSession = Depends(get_db)):
    user_id = await get_user_id(request)
    friend_id = await send_friend_request(session, user_id, body.friend_code)
    if not friend_id:
        raise HTTPException(status_code=400, detail="Could not send request")
    return {"friend_id": friend_id}


@router.post("/remove_friend")
async def remove_friend_endpoint(body: FriendRequest, request: Request, session: AsyncSession = Depends(get_db)):
    user_id = await get_user_id(request)
    result = await session.execute(select(User).where(User.friend_code == body.friend_code))
    friend = result.scalar_one_or_none()
    if not friend:
        raise HTTPException(status_code=404, detail="User not found")
    await remove_friend(session, user_id, friend.id)
    return {"success": True}


@router.post("/accept")
async def accept_request(requester_id: int, request: Request, session: AsyncSession = Depends(get_db)):
    user_id = await get_user_id(request)
    ok = await respond_friend_request(session, user_id, requester_id, accept=True)
    if not ok:
        raise HTTPException(status_code=400, detail="Could not accept")
    return {"success": True}


@router.post("/decline")
async def decline_request(requester_id: int, request: Request, session: AsyncSession = Depends(get_db)):
    user_id = await get_user_id(request)
    ok = await respond_friend_request(session, user_id, requester_id, accept=False)
    if not ok:
        raise HTTPException(status_code=400, detail="Could not decline")
    return {"success": True}

File: js/apps/server/src/routes/friends.ts
import { Router } from 'express';
import { asyncHandler, ApiError } from '../utils/http.js';
import { getUserIdFromAuthHeader } from '../services/authService.js';
import {
  getFriendsData,
  removeFriend,
  respondToFriendRequest,
  sendFriendRequest
} from '../services/friendService.js';

export const friendsRouter = Router();

friendsRouter.use(
  asyncHandler(async (req, _res, next) => {
    await getUserIdFromAuthHeader(req.headers.authorization);
    next();
  })
);

friendsRouter.get(
  '/',
  asyncHandler(async (req, res) => {
    const userId = await getUserIdFromAuthHeader(req.headers.authorization);
    res.json(await getFriendsData(userId));
  })
);

friendsRouter.post(
  '/request/:friendCode',
  asyncHandler(async (req, res) => {
    const userId = await getUserIdFromAuthHeader(req.headers.authorization);
    const friendId = await sendFriendRequest(userId, String(req.params.friendCode));
    res.json({ friendId });
  })
);

friendsRouter.post(
  '/respond/:requesterId',
  asyncHandler(async (req, res) => {
    const userId = await getUserIdFromAuthHeader(req.headers.authorization);
    const action = req.body?.action;
    if (action !== 'accept' && action !== 'decline') {
      throw new ApiError(400, "action must be 'accept' or 'decline'");
    }
    await respondToFriendRequest(userId, String(req.params.requesterId), action);
    res.json({ success: true });
  })
);

friendsRouter.post(
  '/remove/:friendCode',
  asyncHandler(async (req, res) => {
    const userId = await getUserIdFromAuthHeader(req.headers.authorization);
    await removeFriend(userId, String(req.params.friendCode));
    res.json({ success: true });
  })
);
